### ISSUE TITLE
Add custom filters for project jobs

### ISSUE THREAD

**Issue by b-butler** [glotzerlab/signac-flow#267]
### Feature description
Adding custom filters for filtering jobs in the command line and python interface, for status, submission, and running would add flexibility to grouping jobs together. Also, ultimately this would  help with job aggregation for operations (#266, #52). 

### Proposed solution
I think that having a decorator approach like
````python
@FlowProject.filter
def job_filter(job):
     if job.sp.N = 100 and date.is_tuesday():
        return True
````
would be nice. On the command line something like
````bash
python -m project.py status --custom-filter=job_filter
````
would allow the use of the defined filter.

**Comment by bdice** [comment]
I think this is only worth implementing for the job aggregation case, we already have several powerful ways to search that serve most needs. Presently, I would advocate the use of `signac shell` for some of those cases where users might use a filter.

At first I thought we could get by without a decorator at all but that is probably not a good thing for the CLI use case.

I think we already have a "filter" (hence the `-f` option) and would like to avoid overloading terms. Alternative words: selection, selector, tag, category, ...

**Comment by b-butler** [comment]
@bdice, I disagree. I think that there could be cases where a more complicated filter would be warranted for status output or submission. I could see using a label as a filter to prevent having condition checks and label status run on all jobs. For large projects, this could be quite helpful and allows users to not have to go through multiple lines in `signac shell` to do so. Plus there would not be a way to get the pretty output desired for status checks in the shell that I am aware of at least.
